### PR TITLE
Ai opponents

### PR DESCRIPTION
**Pull request by Fueredoriku**
Features:
- Add navmesh to cratertown
- Add navmesh to grand canyon
- Add AI Opponents that follow player with animations  
- Add static persistent data singelton
- Make agents only update paths when their target has moved a certain amount away
- Gun implementation
- Auction implementation 

For later PRs:
- Menu player select with ability to add AIs (will be it's own PR after #90 )
- Track and use winstatistics with persistent data for AI auction meta
- Custom navmeshagent for making AI players use our physics and animations (knockback, leapjumping etc.)
   - This will also allow us to use the fans in grand canyon, which is why I'm not putting more effort into that navmesh atm 
- Gun refinement
   - AIs should have custom behaviors to reload specific augments like DDR and Seeker
   - Ais should try to shoot at players from further away, but aim badly depending on distance (might reuse jiggle vector)
- Smarter movement: ais should strafe back and forth when targeting other players, and try to keep a certain amount of distance away from their target

### CODE REVIEW

**Review by toberge — APPROVED** [review]
Works really well, all jank considered ;)

Main feedback:
- The bots are really good at collecting coins and outbidding you, at least in single player, so we may need to tune them later
- There's some stutter of course
- They sometimes take a while to finish the round after they've offed the player(s)
- and we need to fix rotation and so on
but they don't seem to break the game or anything so let'sa go

File: Assets/Scripts/Auction/BiddingAI.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        currentDestination.TryPlaceBid(playerManager.identity);
+        currentDestination = null;
+    }
+
+    private void Update()
+    {
+        if (!currentDestination)
+            return;
+        animator.SetFloat("Forward", Vector3.Dot(agent.velocity, transform.forward) / agent.speed);
+        animator.SetFloat("Right", Vector3.Dot(agent.velocity, transform.right) / agent.speed);
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Auction/BiddingAI.cs:104]: Ye, not quite sure why that happens, but I'll experiment with setting the forward/right to 0 manually before merging

[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Auction/BiddingAI.cs:104]: Seems to be a navmesh issue, I'll have to fix it later (tm)

File: Assets/Scripts/Gamestate/AIManager.cs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    }
+    public void OnDrawGizmos()
+    {
+        Gizmos.color = Color.red;
+        Gizmos.DrawWireSphere(transform.position, autoAwareRadius);
+    }
+
+    private IEnumerator LookForTargets()
+    {
+        if (!isDead)
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/AIManager.cs:106]: Will fix 👍

File: Assets/Scripts/Gamestate/AIManager.cs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        {
+            Transform closestPlayer = null;
+            float closestDistance = -1f;
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/AIManager.cs:109]: Good catch, I'll set it to the maxArenessRadius 👍

File: Assets/Scripts/Gamestate/AIManager.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                {
+                    // Is the tracked player in front of me? (viewable)
+                    if (Vector3.Dot(transform.forward, targetDirection) < 0)
+                        continue;
+                    // Is there a line of sight to a tracked player?
+                    if (Physics.Raycast(transform.position, targetDirection, hitDistance - 0.1f, ignoreMask))
+                        continue;
+                    // Is there another tracked player who is closer?
+                    if (hitDistance < closestDistance)
+                        continue;
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/AIManager.cs:128]: Yep

File: Assets/Scripts/Gamestate/AIManager.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                DestinationTarget = closestPlayer;
+                // Close enough to shoot!
+                if ((player.transform.position - transform.position).magnitude < 15)
+                    ShootingTarget = player.AiAimSpot;
+            }
+
+            if (closestPlayer == null)
+            {
+                ShootingTarget = null;
+                if (DestinationTarget == null || (!DestinationTarget.gameObject.GetComponent<PlayerManager>() && !DestinationTarget.gameObject.activeInHierarchy))
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/AIManager.cs:142]: Agreed

File: Assets/Scripts/Gamestate/MatchController.cs [hunk 6]
```diff
@@ -114,23 +116,27 @@ void Start()
         // Makes shooting end quickly if testing with 1 player
 #if UNITY_EDITOR
         if (PlayerInputManagerController.Singleton.playerInputs.Count == 1)
-            roundLength = 5f;
+            roundLength = 100f;
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/MatchController.cs:119]: I think we should change it to that when we can select whetever we want AI players or not, I think this time is better now that we can test in training grounds/with starting extension

File: Assets/Scripts/Gamestate/PersistentInfo.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+{
+    public List<CombinationWins> Data = new List<CombinationWins>();
+}
+
+public class PersistentInfo : MonoBehaviour
+{
+    public static PersistentInfo Singleton { get; private set; }
+    [SerializeField]
+    private EditorCombinationWins[] DefaultCombinationStats;
+    public static List<CombinationWins> CombinationStats;
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/PersistentInfo.cs:54]: Eventually™

File: Assets/Scripts/Gamestate/PersistentInfo.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            BinaryFormatter persistentDataFormatter = new BinaryFormatter();
+            FileStream persistentDataStream = File.Open(Application.persistentDataPath + "/PersistentInfo.dat", FileMode.Open);
+            List<CombinationWins> weaponData = new List<CombinationWins>();
+            try
+            {
+                weaponData = ((WinData) persistentDataFormatter.Deserialize(persistentDataStream)).Data;
+            }
+            catch
+            {
+                // File empty or corrupt, reseting to default
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/PersistentInfo.cs:97]: Yup!

File: Assets/Scripts/Gamestate/PlayerFactory.cs [hunk 9]
```diff
@@ -36,31 +44,33 @@ private void Awake()
...
         playerInputManagerController.ChangeInputMaps("Menu");
         InstantiateInputsOnSpawnpoints(InstantiateItemSelectPlayer);
-
     }
 
-    private void InstantiateInputsOnSpawnpoints(Action<InputManager, Transform> instantiate)
+    private List<PlayerManager> InstantiateInputsOnSpawnpoints(Func<InputManager, Transform, PlayerManager> instantiate, Func<int, Transform, AIManager> instantiateAI = null, int aiPlayerCount = 0)
     {
         var shuffledSpawnPoints = new List<Transform>(spawnPoints);
         // Fisher-Yates shuffle
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/PlayerFactory.cs:76]: I'll assign that to the evergrowing "Refactor later" heap

File: Assets/Scripts/Gamestate/MatchController.cs [hunk 10]
```diff
@@ -242,8 +259,12 @@ private IEnumerator DisplayWinScreenAndRestart(PlayerIdentity winner)
 
     public void ReturnToMainMenu()
     {
-        // Update playerInputs in preperation for Menu scene
+        // Update playerInputs / identities in preperation for Menu scene
         PlayerInputManagerController.Singleton.ChangeInputMaps("Menu");
+        // Remove AI identities
+        FindObjectsOfType<PlayerIdentity>()
+            .Where(identity => !identity.GetComponent<InputManager>())
```
[the remark this answers is not in the document]
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/MatchController.cs:266]: Yeah I agree, the AIs need to set properties in their identities before the match starts sooner or later anyway (Like their names), so I'll change this up.

**Pull request merged**